core/lib/zksync_core/src/consistency_checker/mod.rs

```diff
@@ -1,204 +1,300 @@
+        Ok(self
+            .pool
             .access_storage()
-            .await
-            .unwrap()
+            .await?
             .blocks_dal()
             .get_number_of_last_l1_batch_committed_on_eth()
-            .await
-            .unwrap()
-            .unwrap_or(L1BatchNumber(0))
+            .await?)
     }
 
-    pub async fn run(
-        self,
-        stop_receiver: tokio::sync::watch::Receiver<bool>,
-    ) -> anyhow::Result<()> {
-        let mut batch_number: L1BatchNumber = self
+    pub async fn run(mut self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        // It doesn't make sense to start the checker until we have at least one L1 batch with metadata.
+        let earliest_l1_batch_number =
+            wait_for_l1_batch_with_metadata(&self.pool, self.sleep_interval, &mut stop_receiver)
+                .await?;
+
+        let Some(earliest_l1_batch_number) = earliest_l1_batch_number else {
+            return Ok(()); // Stop signal received
+        };
+
+        let last_committed_batch = self
             .last_committed_batch()
-            .await
+            .await?
+            .unwrap_or(earliest_l1_batch_number);
+        let first_batch_to_check: L1BatchNumber = last_committed_batch
             .0
             .saturating_sub(self.max_batches_to_recheck)
-            .max(1)
             .into();
+        // We shouldn't check batches not present in the storage, and skip the genesis batch since
+        // it's not committed on L1.
+        let first_batch_to_check = first_batch_to_check
+            .max(earliest_l1_batch_number)
+            .max(L1BatchNumber(1));
+        tracing::info!(
+            "Last committed L1 batch is #{last_committed_batch}; starting checks from L1 batch #{first_batch_to_check}"
+        );
 
-        tracing::info!("Starting consistency checker from batch {}", batch_number.0);
-
+        let mut batch_number = first_batch_to_check;
         loop {
             if *stop_receiver.borrow() {
                 tracing::info!("Stop signal received, consistency_checker is shutting down");
                 break;
             }
 
-            let metadata = self
-                .db
+            let batch_with_metadata = self
+                .pool
                 .access_storage()
-                .await
-                .unwrap()
+                .await?
                 .blocks_dal()
                 .get_l1_batch_metadata(batch_number)
-                .await
-                .unwrap();
-            let batch_has_metadata = metadata
-                .map(|m| {
-                    m.metadata.bootloader_initial_content_commitment.is_some()
-                        && m.metadata.events_queue_commitment.is_some()
-                })
-                .unwrap_or(false);
+                .await?;
+            let batch_has_metadata = batch_with_metadata.map_or(false, |batch| {
+                // FIXME: how does this work for pre-Boojum batches?
```

Comment:
Unless I'm missing something, pre-Boojum batches will always have `bootloader_initial_content_commitment` and `events_queue_commitment` set to `None`, so the checker will hang up on them.

Reply:
How did it work before Boojum upgrade? Did we remove the corresponding logic? If so, probably it needs to be restored. Although realistically I don't think that it's going to fire, so the priority is probably low, but until we implement a sane policy w.r.t. backward compatibility, I'd always prefer to be consistently backward compatible.

Reply:
Before Boojum, such filtering [was not performed](https://github.com/matter-labs/zksync-era/commit/80273264a9512bc1e6f1d1f4372107f9167260b1#diff-a85ad3ae6a0d60fe1c81964a604bdcecd1eed05a76c87f08f45d992449503db7R163-R177); it was assumed that if an L1 batch has any metadata, it has all of it (i.e., metadata insertion is atomic). I've refactored this logic so that it's easier to reason about.